### PR TITLE
JCLOUDS-871: Azure Copy Blob support

### PR DESCRIPTION
**Pull request by andrewgaul**
Presently lacks support for CopyBlobOptions.  API reference:

http://msdn.microsoft.com/en-us/library/dd894037.aspx

### CODE REVIEW

**Comment by everett-toews** [comment]
Thanks for the pull request but it's release week in jclouds and that means it's time to clean up the PR queue. This PR will be over 6 months old as of April 1. If you intend to continue work on it, please make a comment by April 2. Otherwise it will be closed on April 3.

**Comment by andrewgaul** [comment]
Revived!

File: providers/azureblob/src/main/java/org/jclouds/azureblob/AzureBlobClient.java [hunk 1]
```diff
@@ -254,4 +257,8 @@
     */
    boolean blobExists(String container, String name);
 
+   /**
+    * @throws ContainerNotFoundException if the container is not present.
+    */
+   void copyBlob(URI copySource, String toContainer, String toName, Optional<CopyBlobOptions> options);
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/AzureBlobClient.java]: Changed to use a single non-`Optional` parameter and added `CopyBlobOptions.NONE`.

File: providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.Map;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * @see <a href="http://msdn.microsoft.com/en-us/library/dd894037.aspx" />
+ */
+// TODO:
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java]: Removed/implemented.

File: providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+package org.jclouds.azureblob.options;
+
+import java.util.Map;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * @see <a href="http://msdn.microsoft.com/en-us/library/dd894037.aspx" />
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java]: Done.

File: providers/azureblob/src/main/java/org/jclouds/azureblob/binders/BindAzureCopyOptionsToRequest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      }
+
+      HttpRequest.Builder builder = request.toBuilder();
+      CopyBlobOptions options = optional.get();
+      Optional<Map<String, String>> userMetadata = options.getUserMetadata();
+      if (userMetadata.isPresent()) {
+         for (Map.Entry<String, String> entry : userMetadata.get().entrySet()) {
+            builder.addHeader(AzureStorageHeaders.USER_METADATA_PREFIX + entry.getKey(), entry.getValue());
+         }
+      }
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/binders/BindAzureCopyOptionsToRequest.java]: Correct.

File: providers/azureblob/src/main/java/org/jclouds/azureblob/binders/BindAzureCopyOptionsToRequest.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.jclouds.http.HttpRequest;
+import org.jclouds.rest.Binder;
+
+import com.google.common.base.Optional;
+
+/** Binds options to a copyBlob request. */
+public class BindAzureCopyOptionsToRequest implements Binder {
+   @Override
+   public <R extends HttpRequest> R bindToRequest(R request, Object input) {
+      Optional<CopyBlobOptions> optional = (Optional<CopyBlobOptions>) input;
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/binders/BindAzureCopyOptionsToRequest.java]: Changed.

File: providers/azureblob/src/test/java/org/jclouds/azureblob/AzureBlobClientLiveTest.java [hunk 6]
```diff
@@ -375,4 +380,23 @@ public void testBlockOperations() throws Exception {
...
+
+      // create blob
+      AzureBlob object = getApi().newBlob();
+      object.getProperties().setName("from");
+      object.setPayload(byteSource.read());
+      getApi().putBlob(privateContainer, object);
+
+      // copy blob
+      URI copySource = view.getSigner().signGetBlob(privateContainer, "from").getEndpoint();
+      getApi().copyBlob(copySource, privateContainer, "to", Optional.<CopyBlobOptions>absent());
```
Comment (andrewgaul) [inline providers/azureblob/src/test/java/org/jclouds/azureblob/AzureBlobClientLiveTest.java]: Normal test teardown handles this.

File: providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ */
+// TODO:
+// x-ms-source-if-modified-since
+// x-ms-source-if-unmodified-since
+// x-ms-source-if-match
+// x-ms-source-if-none-match
+public final class CopyBlobOptions {
+   private final Optional<Map<String, String>> userMetadata;
```
Comment (andrewgaul) [inline providers/azureblob/src/main/java/org/jclouds/azureblob/options/CopyBlobOptions.java]: This is an interesting corner case -- Azure does not seem to allow replacing user metadata with an empty map?

**Comment by andrewgaul** [comment]
@demobox Thanks for the feedback and sorry for my delayed responses; please see the latest commit.

**Comment by andrewgaul** [comment]
rebuild please

**Pull request merged**